File: BLSEconomicSurveysAlgorithm.py
# region imports
from AlgorithmImports import *
from QuantConnect.DataSource import *
# endregion


class BLSEconomicSurveysAlgorithm(QCAlgorithm):
    """
    Demonstration algorithm showing how to use BLS Economic Surveys custom datasets
    (CPI, CES, PPI, JOLTS) for signal generation and trading.
    """

    def initialize(self):
        self.set_start_date(2000, 1, 1)
        self.set_end_date(2002, 6, 1)
        self.set_cash(100000)

        # Add a tradeable equity for order execution
        self._spy_symbol = self.add_equity("SPY", Resolution.DAILY).symbol

        # Subscribe to BLS economic surveys (unlinked custom data)
        self._cpi_symbol = self.add_data(BLSEconomicSurveysCpi, "CPI", Resolution.DAILY).symbol
        self._ces_symbol = self.add_data(BLSEconomicSurveysCes, "CES", Resolution.DAILY).symbol
        self._ppi_symbol = self.add_data(BLSEconomicSurveysPpi, "PPI", Resolution.DAILY).symbol

    def on_data(self, slice: Slice):
        # Check for CPI data
        if slice.contains_key(self._cpi_symbol):
            cpi = slice.get(BLSEconomicSurveysCpi, self._cpi_symbol)
            self.log(f"{self.time} - CPI AllItems: {cpi.all_items}, CoreCpi: {cpi.core_cpi}, Energy: {cpi.energy}")

            # Simple signal: if energy CPI is rising faster than core, reduce equity exposure
            if cpi.energy is not None and cpi.core_cpi is not None and cpi.energy > cpi.core_cpi * 1.5:
                if self.portfolio[self._spy_symbol].invested:
                    self.set_holdings(self._spy_symbol, 0.5)

        # Check for employment data
        if slice.contains_key(self._ces_symbol):
            ces = slice.get(BLSEconomicSurveysCes, self._ces_symbol)
            self.log(f"{self.time} - CES TotalNonfarm: {ces.total_nonfarm}, AvgHourlyEarnings: {ces.average_hourly_earnings}")

            # Simple signal: go long when nonfarm payrolls are strong
            if ces.total_nonfarm is not None and ces.total_nonfarm > 130000 and not self.portfolio[self._spy_symbol].invested:
                self.set_holdings(self._spy_symbol, 1)

        # Check for PPI data
        if slice.contains_key(self._ppi_symbol):
            ppi = slice.get(BLSEconomicSurveysPpi, self._ppi_symbol)
            self.log(f"{self.time} - PPI AllCommodities: {ppi.all_commodities}, FarmProducts: {ppi.farm_products}")

File: BLSEconomicSurveysAlgorithm.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using QuantConnect.Data;
using QuantConnect.DataSource;

namespace QuantConnect.Algorithm.CSharp
{
    /// <summary>
    /// Demonstration algorithm showing how to use BLS Economic Surveys custom datasets
    /// (CPI, CES, PPI, JOLTS) for signal generation and trading.
    /// </summary>
    public class BLSEconomicSurveysAlgorithm : QCAlgorithm
    {
        private Symbol _cpiSymbol;
        private Symbol _cesSymbol;
        private Symbol _ppiSymbol;
        private Symbol _spySymbol;

        /// <summary>
        /// Initializes the algorithm with custom data subscriptions.
        /// </summary>
        public override void Initialize()
        {
            SetStartDate(2000, 1, 1);
            SetEndDate(2002, 6, 1);
            SetCash(100000);

            // Add a tradeable equity for order execution
            _spySymbol = AddEquity("SPY", Resolution.Daily).Symbol;

            // Subscribe to BLS economic surveys (unlinked custom data)
            _cpiSymbol = AddData<BLSEconomicSurveysCpi>("CPI", Resolution.Daily).Symbol;
            _cesSymbol = AddData<BLSEconomicSurveysCes>("CES", Resolution.Daily).Symbol;
            _ppiSymbol = AddData<BLSEconomicSurveysPpi>("PPI", Resolution.Daily).Symbol;
        }

        /// <summary>
        /// Handles data events. Uses CPI and employment data to generate trading signals.
        /// </summary>
        /// <param name="slice">The current data slice</param>
        public override void OnData(Slice slice)
        {
            // Check for CPI data
            if (slice.ContainsKey(_cpiSymbol))
            {
                var cpi = slice.Get<BLSEconomicSurveysCpi>(_cpiSymbol);
                Log($"{Time} - CPI AllItems: {cpi.AllItems}, CoreCpi: {cpi.CoreCpi}, Energy: {cpi.Energy}");

                // Simple signal: if energy CPI is rising faster than core, reduce equity exposure
                if (cpi.Energy.HasValue && cpi.CoreCpi.HasValue && cpi.Energy > cpi.CoreCpi * 1.5m)
                {
                    if (Portfolio[_spySymbol].Invested)
                    {
                        SetHoldings(_spySymbol, 0.5);
                    }
                }
            }

            // Check for employment data
            if (slice.ContainsKey(_cesSymbol))
            {
                var ces = slice.Get<BLSEconomicSurveysCes>(_cesSymbol);
                Log($"{Time} - CES TotalNonfarm: {ces.TotalNonfarm}, AvgHourlyEarnings: {ces.AverageHourlyEarnings}");

                // Simple signal: go long when nonfarm payrolls are strong
                if (ces.TotalNonfarm.HasValue && ces.TotalNonfarm > 130000m && !Portfolio[_spySymbol].Invested)
                {
                    SetHoldings(_spySymbol, 1);
                }
            }

            // Check for PPI data
            if (slice.ContainsKey(_ppiSymbol))
            {
                var ppi = slice.Get<BLSEconomicSurveysPpi>(_ppiSymbol);
                Log($"{Time} - PPI AllCommodities: {ppi.AllCommodities}, FarmProducts: {ppi.FarmProducts}");
            }
        }
    }
}
